webExtensionApis/study/schema.yaml

```diff
@@ -0,0 +1,219 @@
+#
+
+# TODO
+# - add json schema types
+# - do we want to expose any of the sampling / hashing stuff?
+# - mention / decide how to handle all
+---
+- namespace: study
+  description: Interface for Shield and Pioneer studies, for use by `background.js` scripts.
+  apiVersion: 5  # for dev use
+  types:   # TODO @glind to actually get this back out of the schemas.json
+  - id: studySetup
+    type: object
+    additionalProperties: {"type": "any" }
+  - id: weightedVariations
+    type: array
+  - id: anEnding
+    type: object
+    additionalProperties: {"type": "any" }
+
+  functions:
+#  ## study lifecycle
+#  - name: setup
+#    type: function
+#    description: |
+#      Configure the study.
+#
+#      - Validate the study configuation, so that all fields needed for study lifecycle telemetry are present.
+#    async: true
+#    parameters:
+#    - name: studySetup
+#      $ref: studySetup
+#      # with a variation included.
+#      # expiration?  -- TODO No, this should be with usual webExt timers.
+
+  - name: setup
+    type: function
+    async: true
+    description: |
+      Attempt an setup/enrollment, with these effects:
+
+      - sets 'studyType' as Shield or Pioneer
+      - Choose variation
+        - using testing.variation OR deterministicVariation
+          for the studyType from `weightedVariations`
+
+      Now Telemetry is sendable.
+
+      - IFF firstRun[1]
+        - set firstrunTimestamp pref value
+        - send 'enter' ping
+        - if `allowEnroll`, send 'install' ping
+        - else endStudy("ineligible") and return
+
+      - Every Run
+        - setActiveExperiment(studyConfig)
+        - monitor shield | pioneer permission endings
+        - suggests alarming if `expire` is set.
+
+      Returns:
+      - info:
+        - state
+        - variation
+        - firstrunTimestamp
+        - endings
+        - weightedVariations
+        - timeUntilExpireAtStartup
+        - activeExperimentName
+
+      Fires:
+      - study:ready
+      - study:endStudy
+
+      Note:
+      1. first run is evaluated based on a pref
+      2. allowEnroll is ONLY used during first run (install)
+
+    returns:  info
+    parameters:
+    - name: studyConfig
+      type: any
+
+  - name: endStudy
+    type: function
+    async: true
+    defaultReturn: {urls: ['url1','url2'], endingName:'some-reason'}
+    description: Optionally opens url, then ends study with pings ending, exit.  Study can only have one ending.  Uninstalls addon?
+    parameters:
+    - name: anEndingName
+      type: ['string', 'null']
+    - name: anEndingObject
+      type: object
+      optional: true
+    # should be anEnding Object
+    returns: directions for what to do next by the addon
+
+  ## study information things
+  - name: info
+    type: function
+    async: true
+    description: current study configuration, including set variation, activeExperimentName
+    defaultReturn: {variation: "styleA"}
+    parameters: []
+
+  - name: dataPermissions
```

Comment:
Attributes related to private browsing, profile age etc, do we leave them outside of utils for now